node/heartbeat/hearbeatMessageInfo.go

```diff
@@ -57,54 +57,60 @@ func newHeartbeatMessageInfo(
 	return hbmi, nil
 }
 
-func (hbmi *heartbeatMessageInfo) updateFields(crtTime time.Time) {
-	validDuration := computeValidDuration(crtTime, hbmi)
-	previousActive := hbmi.isActive && validDuration
-	hbmi.isActive = true
-
-	hbmi.updateTimes(crtTime, previousActive)
-}
-
-func (hbmi *heartbeatMessageInfo) computeActive(crtTime time.Time) {
-	validDuration := computeValidDuration(crtTime, hbmi)
-	hbmi.isActive = hbmi.isActive && validDuration
-
-	hbmi.updateTimes(crtTime, hbmi.isActive)
-}
-
-func (hbmi *heartbeatMessageInfo) updateTimes(crtTime time.Time, previousActive bool) {
+func (hbmi *heartbeatMessageInfo) updateTimes(crtTime time.Time) {
 	if crtTime.Sub(hbmi.genesisTime) < 0 {
 		return
 	}
 	hbmi.updateMaxInactiveTimeDuration(crtTime)
-	hbmi.updateUpAndDownTime(previousActive, crtTime)
-}
-
-func computeValidDuration(crtTime time.Time, hbmi *heartbeatMessageInfo) bool {
-	crtDuration := crtTime.Sub(hbmi.timeStamp)
-	crtDuration = maxDuration(0, crtDuration)
-	validDuration := crtDuration <= hbmi.maxDurationPeerUnresponsive
-	return validDuration
+	hbmi.updateUpAndDownTime(crtTime)
 }
 
 // Will update the total time a node was up and down
-func (hbmi *heartbeatMessageInfo) updateUpAndDownTime(previousActive bool, crtTime time.Time) {
+func (hbmi *heartbeatMessageInfo) updateUpAndDownTime(crtTime time.Time) {
 	if hbmi.lastUptimeDowntime.Sub(hbmi.genesisTime) < 0 {
 		hbmi.lastUptimeDowntime = hbmi.genesisTime
 	}
 
+	if crtTime.Sub(hbmi.timeStamp) < 0 {
+		return
+	}
+
 	lastDuration := crtTime.Sub(hbmi.lastUptimeDowntime)
 	lastDuration = maxDuration(0, lastDuration)
 
-	if previousActive && hbmi.isActive {
-		hbmi.totalUpTime.Duration += lastDuration
-	} else {
-		hbmi.totalDownTime.Duration += lastDuration
+	if lastDuration == 0 {
+		return
 	}
 
+	uptime, downTime := hbmi.computeUptimeDowntime(crtTime, lastDuration)
+
+	hbmi.totalUpTime.Duration += uptime
+	hbmi.totalDownTime.Duration += downTime
+
+	hbmi.isActive = uptime == lastDuration
 	hbmi.lastUptimeDowntime = crtTime
 }
 
+func (hbmi *heartbeatMessageInfo) computeUptimeDowntime(crtTime time.Time, lastDuration time.Duration) (time.Duration, time.Duration) {
+	durationSinceLastHeartbeat := crtTime.Sub(hbmi.timeStamp)
+	insideActiveWindowAfterHeartheat := durationSinceLastHeartbeat <= hbmi.maxDurationPeerUnresponsive
+	noHeartbeatReceived := hbmi.timeStamp == hbmi.genesisTime && !hbmi.isActive
+	outSideActiveWindowAfterHeartbeat := durationSinceLastHeartbeat-lastDuration > hbmi.maxDurationPeerUnresponsive
+
+	uptime := time.Duration(0)
+	downTime := time.Duration(0)
+
+	if noHeartbeatReceived || outSideActiveWindowAfterHeartbeat {
```

Comment:
Might be a good idea to make early exists instead of else if's

Reply:
fixed